File: algorithms/breadth-first_search.java
// The Breadth-First Search (BFS) algorithm is used for traversing or searching tree or graph data structures.
// BFS starts at a specific node of a graph, visits all of its neighbors at the present depth, and then moves on to the nodes at the next depth level.

// BFS can be applied to find the shortest path in a network, such as the shortest route to travel from one city to another in a transportation network.

import java.util.LinkedList;
import java.util.Iterator;

public class BFS {
    static class Graph {
        private int V;
        private LinkedList<Integer>[] adj;

        Graph(int v) {
            V = v;
            adj = new LinkedList[v];
            for (int i = 0; i < v; i++) {
                adj[i] = new LinkedList<>();
            }
        }

        void addEdge(int v, int w) {
            adj[v].add(w);
        }

        void BFS(int s) {
            boolean[] visited = new boolean[V];
            LinkedList<Integer> queue = new LinkedList<>();

            visited[s] = true;
            queue.add(s);

            while (queue.size() != 0) {
                s = queue.poll();
                System.out.print(s + " ");

                Iterator<Integer> i = adj[s].listIterator();
                while (i.hasNext()) {
                    int n = i.next();
                    if (!visited[n]) {
                        visited[n] = true;
                        queue.add(n);
                    }
                }
            }
        }
    }


}

File: algorithms/breadth-first_search.kt
// Breadth-First Search (BFS) is an algorithm for traversing or searching tree or graph data structures.
// It starts at the tree root (or an abtirary node of a graph) and explores the neighbor nodes at the current depth prior to moving on to nodes at the next depth level.

// Imagine you're playing a game where you're in a large maze and you need to find the shortest path to the exit.
// The maze can be visualized as a graph where each room is a node, and each door connecting two rooms is an edge.
// By using BFS, starting from your current room, you'd first explore all the immediate neighboring rooms, then the neighbors of those neighbors, and so on, until you find the exit.
// This ensures you find the shortest path to the exit.

import java.util.LinkedList
import java.util.Queue

class Graph(private val numVertices: Int) {
    private val adjList = MutableList<MutableList<Int>> = ArrayList()

    init {
        for (i in 0 until numVertices) {
            adjList.add(mutableListOf())
        }
    }

    fun addEdge(v: Int, w: Int) {
        adjList[v].add(w)
    }

    fun BFS(startingVertex: Int) {
        val visited = BooleanArray(numVertices)
        val queue: Queue<Int> = LinkedList()

        visited[startingVertex] = true
        queue.add(startingVertex)

        while (queue.isNotEmpty()) {
            val currentIndex = queue.poll()
            print("$currentIndex")

            for (i in adjList[currentVertex]) {
                if (!visited[i]) {
                    visited[i] = true
                    queue.add(i)
                }
            }
        }
    }
}

fun main() {
    val graph = Graph(4)

    graph.addEdge(0, 1)
    graph.addEdge(0, 2)
    graph.addEdge(1, 2)
    graph.addEdge(2, 0)
    graph.addEdge(2, 3)
    graph.addEdge(3, 3)

    println("Breadth First Traversal starting from vertex 2:")
    graph.BFS(2)
}